include/neighbor.h

```diff
@@ -13,57 +13,100 @@ namespace diskann {
   struct Neighbor {
     unsigned id;
     float    distance;
-    bool     flag;
+    bool     checked;
 
     Neighbor() = default;
-    Neighbor(unsigned id, float distance, bool f)
-        : id{id}, distance{distance}, flag(f) {
+    Neighbor(unsigned id, float distance)
+        : id{id}, distance{distance}, checked(false) {
     }
 
     inline bool operator<(const Neighbor &other) const {
       return distance < other.distance ||
-          (distance == other.distance && id < other.id);
+             (distance == other.distance && id < other.id);
     }
 
     inline bool operator==(const Neighbor &other) const {
       return (id == other.id);
     }
   };
 
-  static inline unsigned InsertIntoPool(Neighbor *addr, unsigned K,
-                                        Neighbor nn) {
-    // find the location to insert
-    unsigned left = 0, right = K - 1;
-    if (addr[left].distance > nn.distance) {
-      memmove((char *) &addr[left + 1], &addr[left], K * sizeof(Neighbor));
-      addr[left] = nn;
-      return left;
+  // Invariant: after every `insert` and `pop`, `cur_` points to
+  //            the first Neighbor which is unchecked.
+  class NeighborSet {
+   public:
+    NeighborSet() = default;
+
+    explicit NeighborSet(size_t capacity)
+        : size_(0), capacity_(capacity), data_(capacity_ + 1) {
+    }
+
+    void insert(const Neighbor &nbr) {
+      if (size_ == capacity_ && nbr.distance >= data_[size_ - 1].distance) {
```

Comment:
Neighbor comparisons should use the bool operator< defined on the Neighbor class. This comparison functions includes using the id field to create a stable sort.